src/main/java/bio/terra/common/tracing/TracingHook.java

```diff
@@ -0,0 +1,205 @@
+
+  /** Map of Flight IDs to Flight Spans. */
+  // DO NOT SUBMIT: Consider whether we want to have a FlightHook to match StepHook instead of
+  // storing per Flight data in a global map. Note that the FlightHook would have to be able to make
+  // StepHooks so that the StepHook can refer to data in the FlightHook.
+  private final Map<String, Span> flightSpans = new ConcurrentHashMap<>();
+
+  /** The {@link FlightMap} key for the submission Span's context. */
+  public static final String SUBMISSION_SPAN_CONTEXT_MAP_KEY = "opencensusTracingSpanContext";
+
+  /**
+   * Serialize the current Span's {@link SpanContext}. To be used to store a submission Span for the
+   * Flight with {@link #SUBMISSION_SPAN_CONTEXT_MAP_KEY}.
+   */
+  public static Object serializedCurrentContext() {
+    return encodeContext(Tracing.getTracer().getCurrentSpan().getContext());
+  }
+
+  /** Store the current Span's {@link SpanContext} as the submission Span. */
+  public static void storeCurrentContextAsSubmission(FlightMap inputMap) {
+    inputMap.put(
+        SUBMISSION_SPAN_CONTEXT_MAP_KEY,
+        encodeContext(Tracing.getTracer().getCurrentSpan().getContext()));
+  }
+
+  /**
+   * Creates a Flight Span for the current run of the Flight. Note that the same Flight may be run
+   * multiple times by Stairway.
+   */
+  @Override
+  public HookAction startFlight(FlightContext context) {
+    SpanContext submissionContext = getOrSubmissionContext(context);
+    // Start the Flight Span. We must remember to close the Flight Span at the end of the Flight's
+    // current run.
+    Span flightSpan =
+        Tracing.getTracer()
+            .spanBuilder(FLIGHT_NAME_PREFIX + context.getFlightClassName())
+            .startSpan();
+    flightSpan.addLink(Link.fromSpanContext(submissionContext, Link.Type.PARENT_LINKED_SPAN));
+    flightSpan.putAttribute("stairway/flightId", AttributeValue.stringAttributeValue(context.getFlightId()));
+    flightSpan.putAttribute(
+        "stairway/flightClass", AttributeValue.stringAttributeValue(context.getFlightClassName()));
+    flightSpans.put(context.getFlightId(), flightSpan);
+    return HookAction.CONTINUE;
+  }
+
+  /** Ends the Flight Span for the current run of the Flight. */
+  @Override
+  public HookAction endFlight(FlightContext context) {
+    Span flightSpan = flightSpans.remove(context.getFlightId());
+    // End the Flight Span for the current run.
+    flightSpan.end();
+    return HookAction.CONTINUE;
+  }
+
+  @Override
+  public Optional<StepHook> stepFactory(FlightContext context) {
+    return Optional.of(new TraceStepHook());
+  }
+
+  /** A {@link StepHook} for creating Spans for each Step execution. */
+  private class TraceStepHook implements StepHook {
+    private Span stepSpan = null;
+    private Scope stepScope = null;
+
+    @Override
+    public HookAction startStep(FlightContext flightContext) {
+      Span flightSpan = flightSpans.get(flightContext.getFlightId());
```

Comment:
Couldn't you get this in the constructor?

Reply:
done